### PR TITLE
CopyFilesOverSSHV0 change sequential files sending to continuous queue

### PR DESCRIPTION
**Pull request by ivanduplenskikh**
**Task name**: CopyFilesOverSSHV0

**Description**:
- [x] If contents are set as '**', the **sourceFolder** will be put onto a remote machine with the **uploadDir** method
- [x] If contents will be set, files are sent with continuous queue
- [x] Changed Q to Promise
- [x] Added two task's inputs: **concurrentUploads**, **delayBetweenUploads**

**Documentation changes required:** N

**Added unit tests:** N

**Attached related issue:** [AB#2159883](https://dev.azure.com/mseng/b924d696-3eae-4116-8443-9a18392d8544/_workitems/edit/2159883)

**Checklist**:
- [x] Task version was bumped - please check [instruction](https://github.com/microsoft/azure-pipelines-tasks/tree/master/docs/taskversionbumping.md) how to do it
- [x] Checked that applied changes work as expected

### CODE REVIEW

**Review by DmitriiBobreshev — APPROVED** [review]

File: Tasks/CopyFilesOverSSHV0/sshhelper.ts [hunk 1]
```diff
@@ -103,31 +101,33 @@ export class SshHelper {
...
         }
 
-        const remotePath = path.dirname(dest);
-        try {
-            if (!await this.sftpClient.exists(remotePath)) {
-                await this.sftpClient.mkdir(remotePath, true);
-            }
-        } catch (error) {
-            defer.reject(tl.loc('TargetNotCreated', remotePath));
+        if (this.sshConfig.useFastPut) {
```
Comment (merlynomsft) [inline Tasks/CopyFilesOverSSHV0/sshhelper.ts:108]: missing old code to for exists(path)/mkdir

File: Tasks/CopyFilesOverSSHV0/copyfilesoverssh.ts [hunk 2]
```diff
@@ -183,59 +217,86 @@ async function run() {
...
-                    if (flattenFolders) {
-                        relativePath = path.basename(fileToCopy);
-                    } else {
-                        relativePath = fileToCopy.substring(sourceFolder.length)
-                            .replace(/^\\/g, "")
-                            .replace(/^\//g, "");
-                    }
-                    tl.debug('relativePath = ' + relativePath);
-                    let targetPath = path.posix.join(targetFolder, relativePath);
+        if (contents.length === 1 && contents[0] === "**") {
```
Comment (merlynomsft) [inline Tasks/CopyFilesOverSSHV0/copyfilesoverssh.ts:220]: what does `contents[0] == "**"` mean?  sugget adding a comment

File: Tasks/CopyFilesOverSSHV0/copyfilesoverssh.ts [hunk 3]
```diff
@@ -183,59 +217,86 @@ async function run() {
...
+                // Upload files to remote machine
+                const chunkSize = 10;
+                const chunks = Math.ceil(preparedFiles.length / chunkSize);
+                let successedFiles = 0;
+
+                for (let i = 0; i < chunks; i++) {
+                    // Splice files to copy into chunks
+                    const chunk = preparedFiles.slice(i * chunkSize, i * chunkSize + chunkSize);
+
+                    const results = await Promise.allSettled(chunk.map(async (pathTuple) => {
```
Comment (merlynomsft) [inline Tasks/CopyFilesOverSSHV0/copyfilesoverssh.ts:265]: looks like we queue up ten files to upload at same time and wait for all 10 to finish uploading?  Shouldn't this be a continous queue?

File: _generated/CopyFilesOverSSHV0_Node20/copyfilesoverssh.ts [hunk 4]
```diff
@@ -183,59 +217,86 @@ async function run() {
...
+                        console.log(tl.loc("FolderCreated", foldersPath));
+                    } catch (error) {
+                        throw tl.loc('TargetNotCreated', foldersPath, error);
                     }
+                }
+
+                console.log(tl.loc("FoldersCreated", folderStructure.length));
+
+                // Upload files to remote machine
+                const chunkSize = 10;
```
[the remark this answers is not in the document]
Comment (ivanduplenskikh) [reply _generated/CopyFilesOverSSHV0_Node20/copyfilesoverssh.ts]: The promise mechanism has been redesigned to a queue mechanism.
I've added **CONCURRENT_UPLOADS** and **DELAY_BETWEEN_UPLOADS** variables which are parsed from **process.env**.
These parameters configure the behavior of the queue uploading

File: Tasks/CopyFilesOverSSHV0/copyfilesoverssh.ts (hunk 2 above)
Comment (ivanduplenskikh) [reply Tasks/CopyFilesOverSSHV0/copyfilesoverssh.ts]: This element is added when task's input **contents** has value '**'.
I've added a comment

File: Tasks/CopyFilesOverSSHV0/copyfilesoverssh.ts (hunk 3 above)
Comment (ivanduplenskikh) [reply Tasks/CopyFilesOverSSHV0/copyfilesoverssh.ts]: The mechanism has been redesigned into a continuous queue

File: Tasks/CopyFilesOverSSHV0/sshhelper.ts (hunk 1 above)
Comment (ivanduplenskikh) [reply Tasks/CopyFilesOverSSHV0/sshhelper.ts:108]: Folders creation has been moved to a [stage](https://github.com/microsoft/azure-pipelines-tasks/pull/19798/files/845f325d885d5f6dc5f5df2ccac48bd9502af8c0#diff-00de2e31f64cd05be13d4503059e881bc13bbca5a61d1e75ef90757ab643d004R245) where the **contents** input is not empty.
Now folders creation is based on a result of **getUniqueFolders** function

**Review by DmitriiBobreshev — APPROVED** [review]

**Review by vmapetr — APPROVED** [review]

**Pull request merged**